rfd/0003-plugins-unknown-events.md

```diff
@@ -0,0 +1,143 @@
+state: draft
+---
+
+# RFD 3 - Event handler unknown events
+
+## Required Approvers
+
+* Engineering: @r0mant
+* Product: (@klizhentas || @xinding33)
+
+## What
+
+Customers want to deploy Teleport event handlers that can successfully handle
+all types of events. Currently, Teleport does not provide a way to handle new
+events that are not known to the event handler. This RFD proposes a way to
+handle unknown events when the Teleport event handler version is not up to date
+with the Teleport Auth server version.
+
+## Why
+
+The Teleport Auth server and Teleport event handler are two separate components
+that are deployed independently. This means that the event handler can be
+deployed with a version that is not up to date with the Auth server version.
+This can happen when the Auth server is upgraded but the event handler is not
+upgraded at the same time.
+
+When the event handler is not up to date with the Auth server, it will not be
+able to handle new events that are not known to the event handler. This can
+cause the event handler to crash (previous behavior) or ignores the unknown events.
+
+This is problematic because the event handler is a critical component that
+should be able to handle all types of events. If the event handler is not able
+to handle new events, it can cause the event handler to miss important events
+and drift from the Auth server truth.
+
+The Teleport event handler receives events from the Auth server, serializes them
+using JSON and sends them to the upstream service (e.g. Splunk, Datadog, etc).
+When the event handler is not up to date with the Auth server, the new events
+ids will not be known to the event handler and the protobuf raw cannot be deserialized.
+Without the deserialized event, the event handler cannot serialize the event
+using JSON and send it to the upstream service.
+
+Currently, the event handler ignores unknown events.
+
+## Proposal
+
+The proposal is to have a slower local path to handle unknown events. The slower
+local path consists of having Auth server serving a WASM module that can perform
+the protobuf deserialization and JSON serialization given a raw protobuf event.
+Since the WASM module is served by the Auth server, it will always be up to date
+with the Auth server protobuf version and the event handler will be able to
+download the WASM module each time the Auth server is upgraded/downgraded.
+
+When the event handler receives an unknown event to its protobuf version, the protobuf
+message includes a field called `XXX_unrecognized` that contains the raw slice of bytes
+that wasn't able to be deserialized. If the message with `XXX_unrecognized` is marshaled
+back to protobuf, the `XXX_unrecognized` field will be included in the marshaled message.
+This means that the raw -> unmashaled -> marshaled -> raw path will be the same even if the
+message is unknown to the event handler.
+
+With this in mind, the event handler will be able to send the raw protobuf event to the
```

Comment:
It's possible but not sure if it's the ideal.

While writing this, I spoke with  @hugoShaka about it and he prefers to have a single path - not having a path for known and another for unknown events.  This would simplify the code and allow us to understand bugs.

I changed the RFD so that everything will always hit the WASM routine. 

The current time is:
- `json directly`: 132.699µs
- `via wasm`: 534.154µs -> but involves 2 extra marshals and unmarshals.